core/textile/backup.go

```diff
@@ -0,0 +1,266 @@
+package textile
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/FleekHQ/space-daemon/log"
+
+	"github.com/textileio/go-threads/core/thread"
+)
+
+// backup a single file with an io.Reader provided
+func (tc *textileClient) BackupFileWithReader(ctx context.Context, bucket Bucket, path string, reader io.Reader) (err error) {
+
+	if _, _, err = tc.UploadFileToHub(ctx, bucket, path, reader); err != nil {
+		log.Error(fmt.Sprintf("error backuping up file with reader (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	if _, err = tc.setMirrorFileBackup(ctx, path, bucket.Slug()); err != nil {
+		log.Error(fmt.Sprintf("error setting mirror file as backup (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup a single file with an io.Reader provided
+func (tc *textileClient) backupFile(ctx context.Context, bucket Bucket, path string) error {
+
+	errc := make(chan error, 1)
+	pipeReader, pipeWriter := io.Pipe()
+
+	// go routine for piping
+	go func() {
+		defer close(errc)
+		defer pipeWriter.Close()
+
+		err := bucket.GetFile(ctx, path, pipeWriter)
+		if err != nil {
+			errc <- err
+			log.Error(fmt.Sprintf("error getting file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+			return
+		}
+	}()
+	if err := <-errc; err != nil {
+		return err
+	}
+
+	if err := tc.BackupFileWithReader(ctx, bucket, path, pipeReader); err != nil {
+		log.Error(fmt.Sprintf("error backing up file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup all files in a bucket
+func (tc *textileClient) backupBucketFiles(ctx context.Context, bucket Bucket, path string) (int, error) {
+	var wg sync.WaitGroup
+	var count int
+
+	// XXX: we ignore errc (no atomicity at all) for now but we should return
+	// XXX: the errors, perhaps in a separate async call
+	errc := make(chan error)
+
+	dir, err := bucket.ListDirectory(ctx, path)
+	if err != nil {
+		return 0, err
+	}
+
+	wg.Add(len(dir.Item.Items))
+
+	for _, item := range dir.Item.Items {
+		if item.IsDir {
+			p := strings.Join([]string{path, item.Name}, "/")
+
+			n, err := tc.backupBucketFiles(ctx, bucket, p)
+			if err != nil {
+				return 0, err
+			}
+
+			count += n
+			continue
+		}
+
+		// parallelize the backups
+		go func(path string) {
+			defer wg.Done()
+
+			if err = tc.backupFile(ctx, bucket, path); err != nil {
+				errc <- err
+				return
+			}
+
+			count += 1
+		}(item.Path)
+	}
+
+	wg.Wait()
+
```

Comment:
Why not `<- errc` here? Or is this supposed to be kind of a background process?

Reply:
@dmerrill6 

Thank you for the comment. Mostly because I don't want to overload `errc` (it is for future error handling, not for coordination.)